### ISSUE TITLE
Enhancement: [no-unnecessary-condition] use assignability checks in `checkTypePredicates`

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#11798]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-condition/

### Description

Currently the rule only flags an asserted argument if the type of the variable is _equal_ to the asserted type. I propose to change the logic so that the rule also flags if it's a subtype. 

### Fail

```TypeScript
declare function isStringOrNumber(x: unknown): x is string | number;

declare const s: string;

if (isStringOrNumber(s)) {
}
```

### Pass

```TypeScript
declare function isStringOrNumber(x: unknown): x is string | number;

declare const s: string | number | bigint;

if (isStringOrNumber(s)) {
}
```

### Additional Info

This is split out from https://github.com/typescript-eslint/typescript-eslint/issues/11716#issuecomment-3439853329.

I've begun working on this and there is a bit of nuance beyond simply using `checker.isAssignableTo()` and calling it a day... Examples like the following are troublesome:

```ts
interface Wider {
  a: string;
}

interface Narrower {
  a: string;
  b?: number;
}

declare function isNarrower(x: unknown): x is Narrower;

declare const n: Narrower;
declare const w: Wider;

// These types are mutually assignable:
n satisfies Wider;
w satisfies Narrower;

// Nominally, `isNarrower(w)` seems to be unnecessary, since `w` is 
// already assignable to `Narrower`.
//
// But, using the `isNarrower()` narrowing has a real impact: it allows
// the optional methods in `Narrower` to be invoked on `w`.
if (isNarrower(w)) {
  w.b?.toFixed(); 
}
```

"is assignable to" and "is a subtype of" are not quite same thing in TS, so avoiding false positives on these may require some thought.

cc @ronami
